### PR TITLE
add logfmt promtail stage to be able to extract data from logfmt formatted log

### PR DESCRIPTION
**Pull request by sandeepsukhani**
**What this PR does / why we need it**:
It would be good to have a `logfmt` stage, the same as `json` stage, for extracting data from a log entry. It works the same as `json` stage where the user can specify the mapping for extracted data to define which kvs to extract and set a different key than the one in logs for the extracted data.

**Special notes for your reviewer**:
I would open a flow up PR for adding documentation.

**Checklist**
- [x] Tests updated

### CODE REVIEW

File: clients/pkg/logentry/stages/logfmt.go [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+	"github.com/go-kit/kit/log/level"
+	"github.com/go-logfmt/logfmt"
+	"github.com/mitchellh/mapstructure"
+	"github.com/pkg/errors"
+	"github.com/prometheus/common/model"
+)
+
+// Config Errors
+const (
+	ErrMappingRequired        = "mapping is required"
```
Comment (kavirajk) [inline clients/pkg/logentry/stages/logfmt.go:19]: `ErrLogfmtMappingRequired` = "logfmt mapping is required"?

File: clients/pkg/logentry/stages/logfmt_test.go [hunk 2]
```diff
@@ -0,0 +1,312 @@
...
+
+	for testName, testData := range tests {
+		testData := testData
+
+		t.Run(testName, func(t *testing.T) {
+			t.Parallel()
+
+			pl, err := NewPipeline(util_log.Logger, loadConfig(testData.config), nil, prometheus.DefaultRegisterer)
+			if err != nil {
+				t.Fatal(err)
```
Comment (kavirajk) [inline clients/pkg/logentry/stages/logfmt_test.go:75]: nit: why not `require.NoError()`?

File: clients/pkg/logentry/stages/logfmt_test.go [hunk 3]
```diff
@@ -0,0 +1,312 @@
...
+)
+
+var testLogfmtYamlSingleStageWithoutSource = `
+pipeline_stages:
+- logfmt:
+    mapping:
+      out:  message
+      app:
+      duration:
+      unknown:
```
Comment (kavirajk) [inline clients/pkg/logentry/stages/logfmt_test.go:22]: q: should raise warning if any mapping not found in the parsed `logfmt` line?. Say in this case `unknown`?

Personally, faced these kind of issues if there is some typo in my mapping key.

(ignore me, if we already do that)

File: clients/pkg/logentry/stages/logfmt_test.go [hunk 4]
```diff
@@ -0,0 +1,312 @@
...
+
+var testLogfmtYamlMultiStageWithSource = `
+pipeline_stages:
+- logfmt:
+    mapping:
+      extra:
+- logfmt:
+    mapping:
+      user:
+    source: extra
```
Comment (kavirajk) [inline clients/pkg/logentry/stages/logfmt_test.go:33]: q: So if we have source, then entire `logfmt` parsing is only on that field correct? In this example, we try to parse all the mapping (only `user` here) from `extra` key on the log line.
^^ (this is super nice ❤️ )

How do we delimit the fields inside the source? (here its `=`. But can we have something like `extra="user:foo"` and still extract `user` mapping by having custom delimit `(":")`?

**Review by kavirajk — APPROVED** [review]
Looks super cool! Just few nits!

File: clients/pkg/logentry/stages/logfmt_test.go (hunk 4 above)
Comment (sandeepsukhani) [reply clients/pkg/logentry/stages/logfmt_test.go:33]: There is no way to define a delimiter and logfmt has a standard format to use only `=`. I am using https://github.com/go-logfmt/logfmt for parsing the log message.

File: clients/pkg/logentry/stages/logfmt.go [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+	ErrEmptyLogfmtStageSource = "empty source"
+)
+
+// LogfmtConfig represents a logfmt Stage configuration
+type LogfmtConfig struct {
+	Mapping map[string]string `mapstructure:"mapping"`
+	Source  *string           `mapstructure:"source"`
+}
+
+// validateLogfmtConfig validates a logfmt stage config and returns a map of necessary jmespath expressions.
```
Comment (jeschkies) [inline clients/pkg/logentry/stages/logfmt.go:30]: Does `jmespath` apply here?

File: clients/pkg/logentry/stages/logfmt_test.go [hunk 6]
```diff
@@ -0,0 +1,294 @@
...
+		"successfully run a pipeline with 1 json stage without source": {
+			testLogfmtYamlSingleStageWithoutSource,
+			testLogfmtLogLine,
+			map[string]interface{}{
+				"out":      "this is a log line",
+				"app":      "loki",
+				"duration": "125",
+			},
+		},
+		"successfully run a pipeline with 2 json stages with source": {
```
Comment (jeschkies) [inline clients/pkg/logentry/stages/logfmt_test.go:56]:
```suggestion
		"successfully run a pipeline with 2 logfmt stages with source": {
```

File: clients/pkg/logentry/stages/logfmt_test.go [hunk 7]
```diff
@@ -0,0 +1,294 @@
...
+
+func TestPipeline_Logfmt(t *testing.T) {
+	t.Parallel()
+
+	tests := map[string]struct {
+		config          string
+		entry           string
+		expectedExtract map[string]interface{}
+	}{
+		"successfully run a pipeline with 1 json stage without source": {
```
Comment (jeschkies) [inline clients/pkg/logentry/stages/logfmt_test.go:47]:
```suggestion
		"successfully run a pipeline with 1 logfmt stage without source": {
```

File: clients/pkg/logentry/stages/logfmt_test.go [hunk 8]
```diff
@@ -0,0 +1,294 @@
...
+
+			pl, err := NewPipeline(util_log.Logger, loadConfig(testData.config), nil, prometheus.DefaultRegisterer)
+			assert.NoError(t, err)
+			out := processEntries(pl, newEntry(nil, nil, testData.entry, time.Now()))[0]
+			assert.Equal(t, testData.expectedExtract, out.Extracted)
+		})
+	}
+}
+
+var testLogfmtCfg = `logfmt:
```
Comment (jeschkies) [inline clients/pkg/logentry/stages/logfmt_test.go:80]: I'd move this into the test where it's used. That would make it a little easier to understand the test.

**Review by jeschkies — APPROVED** [review]
Awesome!

File: clients/pkg/logentry/stages/logfmt_test.go (hunk 7 above)
Comment (cyriltovena) [reply clients/pkg/logentry/stages/logfmt_test.go:47]: +1

File: clients/pkg/logentry/stages/logfmt_test.go (hunk 6 above)
Comment (cyriltovena) [reply clients/pkg/logentry/stages/logfmt_test.go:56]: +1

**Review by cyriltovena — APPROVED** [review]
LGTM

Can you take a look at some of those comment before merging.

Thanks

**Pull request merged**